internal/tasks/casign.go

```diff
@@ -0,0 +1,162 @@
+		case AsyncSignOpSignEnrollment:
+			 count, _ = asyncSignEnrollment(ctx, resourceRef, store, callbackManager, log, ca)
+		default:
+			log.Errorf("asyncSign called with unexpected op %s", resourceRef.Op)
+		}
+	}
+	return nil
+}
+
+
+
+func asyncSignEnrollment(ctx context.Context, resourceRef *ResourceReference, dbStore store.Store, callbackManager CallbackManager, log logrus.FieldLogger, ca crypto.CA) (int, error) {
+
+	listParams := store.ListParams{Limit:DefaultBatchSize}
+	selector, err := selector.Parse("!status.certificate")
+
+	if err != nil {
+		log.Info(err)
+		return 0, err
+	}
+	listParams.FieldSelector = selector
+
+	orgId := store.NullOrgId
+	ereqs, err := dbStore.EnrollmentRequest().List(ctx, orgId, listParams)
+	if err != nil {
+		log.Info(err)
+		return 0, err
+	}
+
+	count := 0
+
+	for _, ereq := range ereqs.Items {
+		if ereq.Status != nil && ereq.Status.Certificate == nil && ereq.Status.Conditions != nil && v1alpha1.IsStatusConditionTrue(ereq.Status.Conditions, v1alpha1.EnrollmentRequestApproved) {
+			csr, err := crypto.ParseCSR([]byte(ereq.Spec.Csr))
+			if err := csr.CheckSignature(); err != nil {
+				continue
+			}
+			csr.Subject.CommonName, err = crypto.CNFromDeviceFingerprint(*ereq.Metadata.Name)
+			if err == nil {
+				cert, err := ca.IssueRequestedClientCertificate(csr, DefaultEnrollExpiry)
+				if err == nil {
+					signed := string(cert)
+					ereq.Status.Certificate = &signed
+					_, err = dbStore.EnrollmentRequest().UpdateStatus(ctx, orgId, &ereq)
+					if err != nil {
+						log.Infof("Failed to update enrollment request %s", err)
+					} else {
+						count++
+					}
+				}
+			}
+		}
+	}
+	return count, nil
+
+}
+
+func asyncSignCSR(ctx context.Context, resourceRef *ResourceReference, dbStore store.Store, callbackManager CallbackManager, log logrus.FieldLogger, ca crypto.CA) (int, error) {
+
+
+	listParams := store.ListParams{Limit:DefaultBatchSize}
+	selector, err := selector.Parse("!status.certificate")
+
+	if err != nil {
+		log.Info(err)
+		return 0, err
+	}
+	listParams.FieldSelector = selector
+
+	orgId := store.NullOrgId
+	ereqs, err := dbStore.CertificateSigningRequest().List(ctx, orgId, listParams)
+	if err != nil {
+		log.Info(err)
+		return 0, err
+	}
+
+	count := 0
+
+	for _, ereq := range ereqs.Items {
+		if ereq.Status == nil {
+			ereq.Status = &v1alpha1.CertificateSigningRequestStatus{}
+		}
+		if ereq.Status.Certificate == nil {
+			csr, err := crypto.ParseCSR([]byte(ereq.Spec.Request))
+			if err != nil {
+				continue
+			}
+			if err := csr.CheckSignature(); err != nil {
+				continue
+			}
+			u := csr.Subject.CommonName
+			if u == "" {
+				u = uuid.NewString()
+			}
+			csr.Subject.CommonName, err = crypto.BootstrapCNFromName(u)
+			if err != nil {
+				continue
+			}
+			expiry := DefaultCSRExpiry
+			if ereq.Spec.ExpirationSeconds != nil {
+				expiry = int(*ereq.Spec.ExpirationSeconds)
+			}
+			cert, err := ca.IssueRequestedClientCertificate(csr, expiry)
```

Comment:
should we log/return this err?

Reply:
We cannot return it. At all. We can log it, but that will be buried in the log. 

This is a common problem for all async jobs which I raised on the tech call last week. We have to record/log the state of async processing somewhere where it can be accessed by an API. Either tag an array of records on Status or start a proper database based event log.